runtime/storage.go

```diff
@@ -283,117 +177,143 @@ func (s *Storage) recordContractUpdate(
+	if storable == nil {
+		value = interpreter.NilValue{}
+	} else {
+		contractValue := interpreter.StoredValue(storable, s)
+		value = interpreter.NewSomeValueNonCopying(contractValue)
+	}
 
-		storageKey := entry.StorageKey
-		storable := entry.Storable
+	storageMap.WriteValue(inter, key.Key, value)
+}
 
-		address := storageKey.Address
+type write struct {
+	storageKey   interpreter.StorageKey
+	storageIndex atree.StorageIndex
+}
 
-		// If the account storage change is a contract update,
-		// and it is overwriting an existing contract value,
-		// it must be properly removed first:
-		// The removal did not occur during execution,
-		// because contract updates are deferred to the commit
+func sortWrites(writes []write) {
+	sort.Slice(writes, func(i, j int) bool {
+		a := writes[i].storageKey
+		b := writes[j].storageKey
+		return a.IsLess(b)
+	})
+}
 
-		if entry.IsContractUpdate {
-			existingStorable := s.readStorable(storageKey)
-			if existingStorable != nil {
-				interpreter.StoredValue(existingStorable, s).
-					DeepRemove(inter)
-				inter.RemoveReferencedSlab(existingStorable)
-			}
-		}
+// Commit serializes/saves all values in the readCache in storage (through the runtime interface).
+//
+func (s *Storage) Commit(inter *interpreter.Interpreter, commitContractUpdates bool) error {
 
-		var encoded []byte
+	if commitContractUpdates {
+		s.commitContractUpdates(inter)
+	}
 
-		if storable != nil {
-			var err error
+	var writes []write
 
-			var buf bytes.Buffer
-			encoder := atree.NewEncoder(&buf, interpreter.CBOREncMode)
+	writeCount := len(s.writes)
+	if writeCount > 0 {
+		writes = make([]write, 0, writeCount)
+	}
 
-			s.reportMetric(
-				func() {
-					err = storable.Encode(encoder)
-				},
-				func(metrics Metrics, duration time.Duration) {
-					metrics.ValueEncoded(duration)
-				},
-			)
-			if err != nil {
-				return err
-			}
+	// NOTE: ranging over maps is safe (deterministic),
+	// if it is side effect free and the keys are sorted afterwards
+
+	for storageKey, storageIndex := range s.writes { //nolint:maprangecheck
+		writes = append(
+			writes,
+			write{
+				storageKey:   storageKey,
+				storageIndex: storageIndex,
+			},
+		)
+	}
 
-			err = encoder.CBOR.Flush()
-			if err != nil {
-				return err
-			}
+	// Sort the writes by storage key in lexicographic order
+	if writeCount > 1 {
+		sortWrites(writes)
+	}
 
-			encoded = buf.Bytes()
-		}
+	// Write account storage entries in order
+
+	// NOTE: Important: do not use a for-range loop,
+	// as the introduced variable will be overridden on each loop iteration,
+	// leading to the slices created in the loop body being backed by the same data
+	for i := 0; i < len(writes); i++ {
+		write := writes[i]
 
 		var err error
 		wrapPanic(func() {
 			err = s.Ledger.SetValue(
-				address[:],
-				[]byte(storageKey.Key),
-				encoded,
+				write.storageKey.Address[:],
+				[]byte(write.storageKey.Key),
+				write.storageIndex[:],
```

Comment:
What do you think about removing `writes` cache entry after `Ledger.SetValue` is returned?  I think get storage used function can call `Commit` multiple times and cause the same value to be written to ledger more than once.

Reply:
I guess that makes sense, I added this in 61dee3f69f21b2fd695aef133a688a965393cc6c